### ISSUE TITLE
[Bug]: silent miscompile — `(vec_bool != scalar_true)` folds to all-true regardless of scalar source (`tl.load` or literal); floor-mod sign correction over-fires for negative divisors

### ISSUE THREAD

**Issue by gavinwang269** [triton-lang/triton-ascend#2290]
## 问题现象

triton-ascend 3.6.0 在**纯 Triton 内核**（不经 torch.compile/Inductor）里对「向量 bool 谓词 `!=` 标量 bool 谓词」的组合存在**静默误编译**：只要**标量侧为真**，组合结果就被折叠为**全真**——与标量的来源（`tl.load` 载入的 0-dim 值、还是 `tl.constexpr` 字面量）**无关**，与 dtype 无关（int32/int64/float16/float32 同型）。内核不报任何错，直接产出错误数值。

标准 C-rem→floor-mod 符号修正链 `r = where(((r<0)!=(b<0)) & (r!=0), r+b, r)` 因此对**所有**负除数整体失效：除数 `b<0` 时标量谓词 `(b<0)` 为真 → 折叠全真 → 每个 `r != 0` 的 lane 都多做一次 `+b`（系统性过冲一个除数量），而 `b>0` 时（标量谓词为假）完全正确。

复现 A（确定性网格输入）实测形态：

- `b = -4`：`tl.load` 谓词版与字面量谓词版**同样错**，各 1536/4096 lanes，错误 lane 恰为 `a<0 且 a%4≠0` 的 48 个值（每值重复 32 次），错值一律 = 正确 floor 结果 `+ b`（单次过冲）；
- `b = +4`：两版全对（0/4096）；
- 同一内核仅把 `a`/`b` 换为 `int64` / `float16` / `float32`（`b_val=±4.0`）：四格计数与 int32 逐位一致；
- 连续随机输入（`randn*30`，float16/float32）下错误率 ≈50%，恰为负数占比——同一折叠形态的随机版投影。

把错误**隔离到 `!=` 组合本身**的三组对照（复现 B）：

- 同一真值标量改为**向量**（splat `-4`）参与 `!=`：正确（2560/4096 true，逐 lane 期望值）；
- 同一标量改写 xor 的**展开式** `((r<0)&(b>=0)) | ((r>=0)&(b<0))`：正确；
- 标量侧为**假**（`b=+4`）的 `!=`：正确。
- 其余算术无罪：`a % b` 本身在全部实测 dtype 与除数（±4、±INT_MAX、INT_MIN、2^30/2^62 等，`tl.load` 标量与向量除数）下 **100% 等于** eager `torch.fmod`（C 式截断取余，即文档语义）。

即：**唯一触发条件是 `!=` 右操作数为「值为真的标量 bool」**，与来源、dtype、其余数值无关。

## 环境

| 组件 | 版本 |
|---|---|
| 硬件 | Ascend 910B4 |
| triton-ascend | 3.6.0（`triton.__version__` 报 3.5.0） |
| PyTorch | 2.14.0a0（dev） |
| torch_npu | 2.14.0 |
| CANN | 9.1.0 GA |
| Python | 3.12 |

两个复现均只需 triton-ascend + eager torch（不依赖 torch.compile/`TORCHINDUCTOR_NPU_BACKEND`）。前置：已加载 CANN 环境（`source <CANN 安装目录>/ascend-toolkit/set_env.sh`）并导入 `torch_npu`。

## 复现 A：纯 Triton floor-mod 内核（主证据，确定性计数）

```python
# repro_floor_mod_fold.py
import torch
import torch_npu  # noqa: F401  注册 NPU device
import triton
import triton.language as tl

@triton.jit
def floor_mod_kernel(a_ptr, b_ptr, out_ptr, N,
                     B_CONST: tl.constexpr, USE_LITERAL: tl.constexpr,
                     BLOCK: tl.constexpr):
    offs = tl.program_id(0) * BLOCK + tl.arange(0, BLOCK)
    mask = offs < N
    a = tl.load(a_ptr + offs, mask=mask)
    b = tl.load(b_ptr)                                   # 标量除数（0-dim tensor）
    r = a % b                                            # C-style remainder（实测无罪）
    if USE_LITERAL:
        neg = ((r < 0) != (B_CONST < 0)) & (r != 0)      # 字面量来源的标量谓词
    else:
        neg = ((r < 0) != (b < 0)) & (r != 0)            # tl.load 来源的标量谓词
    out = tl.where(neg, r + b, r)                        # 标准 C-rem→floor-mod 修正
    tl.store(out_ptr + offs, out, mask=mask)

N = 4096
a_cpu = torch.arange(-64, 64, dtype=torch.int32).repeat(32)   # 每值恰好重复 32 次
for b_val in (-4, 4):
    a = a_cpu.npu()
    b = torch.tensor(b_val, device="npu", dtype=torch.int32)
    ref = torch.remainder(a, b)                          # eager 语义基准
    for use_lit in (False, True):
        out = torch.empty_like(a)
        floor_mod_kernel[(N // 1024,)](
            a, b, out, N, B_CONST=b_val, USE_LITERAL=use_lit, BLOCK=1024)
        tag = "literal-predicate" if use_lit else "tl.load-predicate"
        print(f"b={b_val:+d}  {tag:17s}: {int((out != ref).sum())}/{N} lanes wrong")
```

预期（逐 lane 语义）四格全 `0/4096`；实际逐字输出：

```
b=-4  tl.load-predicate : 1536/4096 lanes wrong
b=-4  literal-predicate : 1536/4096 lanes wrong
b=+4  tl.load-predicate : 0/4096 lanes wrong
b=+4  literal-predicate : 0/4096 lanes wrong
```

把 `a_cpu`/`b`/`ref` 的 dtype 换为 `torch.int64`、`torch.float16`、`torch.float32`（`b_val=±4.0` 对应浮点张量）重跑，四格计数不变。

## 复现 B：把误折叠隔离到「`!=` × 真值标量」

```python
# repro_xor_forms.py
import torch
import torch_npu  # noqa: F401
import triton
import triton.language as tl

@triton.jit
def xor_forms(a_ptr, b_ptr, bvec_ptr, o1_ptr, o2_ptr, o3_ptr, N, BLOCK: tl.constexpr):
    offs = tl.program_id(0) * BLOCK + tl.arange(0, BLOCK)
    mask = offs < N
    a = tl.load(a_ptr + offs, mask=mask)
    b = tl.load(b_ptr)                                   # 标量除数
    bv = tl.load(bvec_ptr + offs, mask=mask)             # 向量除数（全元素同 b）
    r = a % b
    o1 = ((r < 0) != (b < 0)).to(a.dtype)                            # 向量 != 标量
    o2 = ((r < 0) != (bv < 0)).to(a.dtype)                            # 向量 != 向量
    o3 = (((r < 0) & (b >= 0)) | ((r >= 0) & (b < 0))).to(a.dtype)    # 同式的展开写法
    tl.store(o1_ptr + offs, o1, mask=mask)
    tl.store(o2_ptr + offs, o2, mask=mask)
    tl.store(o3_ptr + offs, o3, mask=mask)

N = 4096
a = torch.arange(-64, 64, dtype=torch.int32).repeat(32).npu()
for b_val in (-4, 4):
    b = torch.tensor(b_val, device="npu", dtype=torch.int32)
    bvec = torch.full_like(a, b_val)
    o = [torch.empty_like(a) for _ in range(3)]
    for blk in range(N // 1024):
        xor_forms[(blk,)](a, b, bvec, *o, N, BLOCK=1024)   # 注：grid 逐块需 offs 平移，见下
    r = torch.fmod(a, b)                                   # 实测 100% 可信（C 式截断取余）
    exp = (r < 0) != (b < 0)                               # 逐 lane 真值基准（host 侧语义）
    for name, t in zip(("!= scalar  ", "!= vector ", "expanded  "), o):
        print(f"b={b_val:+d} {name}: true-count {int(t.sum())}/{N}  (expected {int(exp.sum())}/{N})")
```

预期（逐 lane 语义）三式在两个 `b` 号下都与 host 基准 `(r<0)!=(b<0)` 计数一致；实际逐字输出：

```
b=-4 != scalar  : true-count 4096/4096  (expected 2560/4096)   ← 折叠为全真
b=-4 != vector  : true-count 2560/4096  (expected 2560/4096)
b=-4 expanded   : true-count 2560/4096  (expected 2560/4096)
b=+4 != scalar  : true-count 1536/4096  (expected 1536/4096)
b=+4 != vector  : true-count 1536/4096  (expected 1536/4096)
b=+4 expanded   : true-count 1536/4096  (expected 1536/4096)
```

## 期望 vs 实际

- **期望**：`!=` 对向量×标量按逐 lane 语义求值（标量广播），三种写法（`!=` 标量 / `!=` 向量 / 展开式）与两个 `b` 号下的 6 格计数全部一致、与 host 侧 `(r<0)!=(b<0)` 真值逐位一致；复现 A 四格 0 错。
- **实际**：仅「标量侧为真」的 `!=` 组合被折叠为全真（复现 B `b=-4` 第一行 4096/4096；复现 A `b=-4` 两格各 1536 lanes），无编译错误、无运行时报错；`b=+4` 与其余写法全部正确。

## 相邻缺陷（同环境实测确认，均**不**属本单范围）

- `float64` 的 `a % b`：编译期 `LLVM ERROR: unsupported type cast.`（bishengir 崩溃，非静默错值）——另行跟踪。
- 整型 `//` 配 `tl.constexpr` **字面量**除数时对大 |a| 静默错值，而同一除数的 `%` 正确（div/rem 降级不一致）：`2147483646 // 3 → 715827904`（期望 − 向零 715827882 / 向下取整 715827883，两义均不符）、`2147483646 // 65536 → 32768`（期望 32767）、`2147483646 // 2147483647 → 1`（期望 0）、`-2147483648 // 3 → -715827904`；`tl.load` 标量除数的 `//` 实测无误。形态与本单不同，另行跟踪。
- f16/f32 的 0-dim / 双 0-dim `torch.floor_divide` 编译期 `[ConvertTritonIRToLinalgIR]` 崩溃（f64 正常）：属配套编译崩溃单范围，与本单互斥。
- torch.compile(triton_experimental) `unravel_index` 0-dim 大除数形（如 `unravel_index(tensor(-1), (1, 2**63-1))` 编译输出第二元素为 0）：读生成码系 int64 常量被物化为 `tl.full([...], 9223372036854775807, tl.int32)`（截断成 −1，随后 `x % -1 ≡ 0`）——codegen 常量 dtype 问题，根因与本单折叠无关，另行跟踪。另：早期形态 `arange(0,120,20)` × `(2,3,4,5)` 复测编译输出与 eager 完全一致，不再作为本单证据。

## 与 torch_npu 侧规避的关系

- 配套工单 Ascend/pytorch#4942 / 规避 PR Ascend/pytorch#46602：将 `aten.remainder` 移出 TE 后端 `GENERATE_LIST`、转 eager fallback。本次复测中该环境编译路径的 `remainder`（0-dim 除数，含负值）输出已正确（回退生效），**但本单折叠形态在纯 Triton 内核里独立存在**（复现 A/B），任何手写或未来代码生成路径只要产出 `(vec != 真标量)` 形态即中招，编译器修复仍是唯一收口。
- 上述规避 PR 注释里若记录了「`!=` 仅当标量侧来自 `tl.load` 才误折、字面量侧正确」的早期形态描述，请按本单复现 B 修正为：**折叠与标量来源无关，触发条件是标量谓词为真**。
- 本缺陷静默（错误数值直接进下游），对生产图的实际影响面大于有报错的编译失败，请优先排查布尔 `!=`/xor 与标量广播相关的折叠 pass。

## 附录：观测清单（910B4 + 上表版本，全部本轮实测）

| 形态 | 结果 |
|---|---|
| floor-mod 内核 `b=-4`（`tl.load` 标量谓词），int32 | 1536/4096 错；`neg` 真值计数 3072（期望 1536）；错 lane 一律 = floor+b |
| floor-mod 内核 `b=-4`（字面量标量谓词），int32 | 与上行逐 lane 相同 |
| floor-mod 内核 `b=+4`（两种来源） | 0 错 |
| 同上换 int64 / float16 / float32（b=±4.0） | 计数与 int32 完全一致 |
| 随机 `randn*30` float16/float32，`b=-4.0` | 两来源均 ≈50% 错（负数占比，同形态投影） |
| 裸 `(r<0)!=(b<0)`，标量真 / 标量假 / 向量 / 展开式 | 4096全真（错） / 正确 / 正确 / 正确（复现 B） |
| `a % b` 全 dtype 全除数对照（±4、±INT_MAX、INT_MIN、2^30、2^62，含向量除数） | 100% == `torch.fmod`（C 式截断取余，语义正确） |
| eager / 纯 host 基准、`torch.remainder` 对照 | 全部正确（对照组） |

**Comment by hongziqi** [comment]
Thanks for your feedback. We will arrange relevant engineers to investigate this issue later.

**Comment by liusongshuo** [comment]
Thanks for the detailed report and the minimal repros. We reproduced the silent miscompile on the **official Triton-Ascend 3.6.0 release** package (910B4 / CANN 9.1.0): with a negative divisor, the predicate `(r < 0) != (b < 0)` folds to all-true, so your floor-mod kernel produces `r + b` on every lane where `r != 0`:

```text
b=-4  tl.load-predicate : 1536/4096 lanes wrong
b=-4  literal-predicate : 1536/4096 lanes wrong
b=+4  tl.load-predicate : 0/4096 lanes wrong
b=+4  literal-predicate : 0/4096 lanes wrong
```

As you identified, the fold depends only on the scalar side being true — the scalar's source (`tl.load` or literal) is irrelevant, while a vector operand is unaffected.

## Root cause

The Ascend adapter IR for this kernel is fine. The miscompile happens later inside **NPUIR**: `hfusion-normalize-ops` legalizes `ne` on `i1` as an `f16` equality comparison, but the two sides are converted with opposite sign conventions — the vector side (`cast_signed` → `vcast`) maps true to `+1.0` while the scalar side (`arith.sitofp`) maps true to `-1.0`. The `veq` is therefore always false and the folded `vnot` always true.

## Fix / recommendation

This has been fixed in an updated `bishengir-compile` bundled with the **3.6.0 daily** wheels. The official 3.6.0 release package has not yet picked up that BiShengIR update, so the same repro still miscompiles there.

Since the next official release containing this fix has not been published yet, please use the daily build for now — the fix will also be included in the upcoming official release.

Please install a **3.6.0 daily** build (2026-09-22) instead of the older release wheel:

- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp310-cp310-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp310-cp310-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp311-cp311-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp311-cp311-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp312-cp312-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp312-cp312-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp313-cp313-manylinux_2_27_aarch64.manylinux_2_28_aarch64.whl
- https://ascend-triton-open.obs.cn-north-4.myhuaweicloud.com/triton-ascend/20260922220228/triton_ascend-3.6.0%2Bdev20260922220228-cp313-cp313-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl

We re-verified your floor-mod repro (`repro_floor_mod.py`) verbatim on that daily package (`3.6.0+dev20260922220228`, 910B4) — all four cells are now correct:

```text
b=-4  tl.load-predicate : 0/4096 lanes wrong
b=-4  literal-predicate : 0/4096 lanes wrong
b=+4  tl.load-predicate : 0/4096 lanes wrong
b=+4  literal-predicate : 0/4096 lanes wrong
```
